flow_judge/models/adapters/baseten/adapter.py

```diff
@@ -0,0 +1,169 @@
+        parsed_resp = self._make_request(request_body)
+
+        try:
+            message = parsed_resp["choices"][0]["message"]["content"].strip()
+            return message
+        except Exception as e:
+            # TODO: Log warning here (?)
+            return ""
+        
+    def fetch_batched_response(self, request_messages: list[Dict[str, Any]]) -> list[str]:
+        outputs = []
+        for message in request_messages:
+            request_body = {"messages": message}
+            parsed_resp = self._make_request(request_body)
+            outputs.append(parsed_resp["choices"][0]["message"]["content"].strip())
+        return outputs
+
+class AsyncBasetenAPIAdapter(BaseAPIAdapter):
+    """Async webhook requests for the Baseten remote model"""
+    def __init__(self, baseten_model_id: str, webhook_proxy_url: str):
+        super().__init__(f"https://model-{baseten_model_id}.api.baseten.co/production")
+
+        self.baseten_model_id = baseten_model_id
+        self.webhook_proxy_url = webhook_proxy_url
+
+        try:
+            self.baseten_api_key = os.environ["BASETEN_API_KEY"]
+        except KeyError:
+            raise ValueError("BASETEN_API_KEY is not provided in the environment.")        
+
+    def _make_request(self, request_messages: Dict[str, Any]) -> str:
+        model_input = {"messages": request_messages}
+        resp = requests.post(
+            url=self.base_url + "/async_predict",
+            headers={"Authorization": f"Api-Key {self.baseten_api_key}"},
+            json={
+                "webhook_endpoint": self.webhook_proxy_url + "/webhook",
+                "model_input": model_input
+            }
+        )
+        return resp.json()["request_id"]
+
+    def fetch_response(self, request_messages: Dict[str, Any]) -> str:
+        request_id = self._make_request(request_messages)
+        stream = requests.get(f"{self.webhook_proxy_url}/listen/{request_id}", stream=True)
+        message = ""
+        for chunk in stream.iter_content(chunk_size=None):
+            if not chunk:
+                break
+            elif chunk.decode() == "data: keep-alive\n\n":
+                continue
+            else:
+                decoded_chunk = chunk.decode()
+                data_and_eot = decoded_chunk.split("\n\n")
+                data_chunk = data_and_eot[0]
+                resp_str = data_chunk.split("data: ")[1] if data_chunk.startswith("data: ") else data_chunk
+
+                try:
+                    resp = json.loads(resp_str)
+                    message = resp["data"]["choices"][0]["message"]["content"].strip()
+                except Exception as e:
+                    # raise? warn?
+                    continue
+
+                if len(data_and_eot) > 1 and "\n\ndata: eot" in data_and_eot[1]:
+                    stream.close()
+
+        return message
+
+    def fetch_batched_response(self, request_messages: list[Dict[str, Any]]) -> list[str]:
```

Comment:
@alexwegrzyn @minaamshahid 

This is the problem function yes?

Some considerations I have:
1. Why couple to the same function both the fire-and-forget requests to the baseten async endpoint AND then wait for the responses in the same place?
2. Is this function synchronous or asychronous?

Reply:
![image](https://github.com/user-attachments/assets/67e7f072-30ba-434e-b71c-793637440b6a)

Reply:
I guess normally asyncio is used, like Claude would spit out something like this

```
class AsyncBasetenAPIAdapter(BaseAPIAdapter):
    """Async webhook requests for the Baseten remote model"""
    def __init__(self, baseten_model_id: str, webhook_proxy_url: str):
        super().__init__(f"https://model-{baseten_model_id}.api.baseten.co/production")
        self.baseten_model_id = baseten_model_id
        self.webhook_proxy_url = webhook_proxy_url
        try:
            self.baseten_api_key = os.environ["BASETEN_API_KEY"]
        except KeyError:
            raise ValueError("BASETEN_API_KEY is not provided in the environment.")

    async def _make_request(self, request_messages: Dict[str, Any]) -> str:
        model_input = {"messages": request_messages}
        async with aiohttp.ClientSession() as session:
            async with session.post(
                url=self.base_url + "/async_predict",
                headers={"Authorization": f"Api-Key {self.baseten_api_key}"},
                json={
                    "webhook_endpoint": self.webhook_proxy_url + "/webhook",
                    "model_input": model_input
                }
            ) as response:
                resp_json = await response.json()
                return resp_json["request_id"]

    async def _fetch_stream(self, request_id: str) -> str:
        async with aiohttp.ClientSession() as session:
            async with session.get(f"{self.webhook_proxy_url}/listen/{request_id}") as response:
                message = ""
                async for chunk in response.content.iter_any():
                    decoded_chunk = chunk.decode()
                    if decoded_chunk == "data: keep-alive\n\n":
                        continue
                    data_and_eot = decoded_chunk.split("\n\n")
                    data_chunk = data_and_eot[0]
                    resp_str = data_chunk.split("data: ")[1] if data_chunk.startswith("data: ") else data_chunk
                    try:
                        resp = json.loads(resp_str)
                        message = resp["data"]["choices"][0]["message"]["content"].strip()
                    except json.JSONDecodeError as e:
                        logging.warning(f"Failed to parse JSON for request_id {request_id}: {e}")
                        continue
                    if len(data_and_eot) > 1 and "\n\ndata: eot" in data_and_eot[1]:
                        break
                return message

    async def fetch_response(self, request_messages: Dict[str, Any]) -> str:
        request_id = await self._make_request(request_messages)
        return await asyncio.wait_for(self._fetch_stream(request_id), timeout=120)  # 2 minutes timeout

    async def fetch_batched_response(self, request_messages: List[Dict[str, Any]]) -> List[str]:
        request_ids = await asyncio.gather(*[self._make_request(message) for message in request_messages])
        tasks = [self._fetch_stream(request_id) for request_id in request_ids]
        results = await asyncio.gather(*[asyncio.wait_for(task, timeout=120) for task in tasks], return_exceptions=True)
        return [result if not isinstance(result, Exception) else "" for result in results]

    def fetch_response_sync(self, request_messages: Dict[str, Any]) -> str:
        return asyncio.run(self.fetch_response(request_messages))

    def fetch_batched_response_sync(self, request_messages: List[Dict[str, Any]]) -> List[str]:
        return asyncio.run(self.fetch_batched_response(request_messages))
```

Reply:
Having the functions be async would already decouple sending and waiting.